src/librustc/ty/sty.rs

```diff
@@ -556,9 +556,39 @@ pub struct ProjectionTy<'tcx> {
     /// The trait reference `T as Trait<..>`.
     pub trait_ref: ty::TraitRef<'tcx>,
 
-    /// The name `N` of the associated type.
-    pub item_name: Name,
+    /// The DefId of the the associated type `N`.
+    ///
+    /// Note that this is not the DefId of the TraitRef containing this
+    /// associated type, which is in tcx.associated_item(item_def_id).container.
+    pub item_def_id: DefId,
 }
+
+impl<'a, 'tcx> ProjectionTy<'tcx> {
+    pub fn from_ref_and_name(
```

Comment:
Maybe add a comment here?

"Construct a `ProjectionTy` by searching the trait from `trait_ref` for the associated item named `item_name`."